### ISSUE TITLE
Unable to use clang header with GCC <9.3 with `-Werror`

### ISSUE THREAD

**Issue by jabraham17** [llvm/llvm-project#109192]
Due to a bug in GCC, any compilation with GCC older than 9.3 that uses `clang/Basic/PointerAuthOptions.h` throws a warning:

"‘clang::PointerAuthSchema::DiscriminationKind’ is too small to hold all values of ‘enum class clang::PointerAuthSchema::Discrimination’"

The [GCC Bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=51242) was fixed in 9.3 (also 8.4), the following shows the issue with various compilers: https://godbolt.org/z/Ko48cbj1T. 

While this is only a warning, it is now impossible to compile code with `-Werror`, because there is no separate warning flag for this warning. So the warning cannot be turned off and it cannot be marked `-Wno-error=...`.

This is a regression in clang 19, caused by https://github.com/llvm/llvm-project/pull/93906.

The following patch can be applied to `clang/Basic/PointerAuthOptions.h` and is sufficient to workaround the issue, and I strongly feel this should be included in a point release

```patch
@@ -63,17 +63,17 @@
   };
 
 private:
-  Kind TheKind : 2;
+  unsigned TheKind : 2;
   unsigned IsAddressDiscriminated : 1;
   unsigned IsIsaPointer : 1;
   unsigned AuthenticatesNullValues : 1;
-  PointerAuthenticationMode SelectedAuthenticationMode : 2;
-  Discrimination DiscriminationKind : 2;
+  unsigned SelectedAuthenticationMode : 2;
+  unsigned DiscriminationKind : 2;
   unsigned Key : 2;
   unsigned ConstantDiscriminator : 16;
 
 public:
-  PointerAuthSchema() : TheKind(Kind::None) {}
+  PointerAuthSchema() : TheKind(llvm::to_underlying(Kind::None)) {}
 
   PointerAuthSchema(
       ARM8_3Key Key, bool IsAddressDiscriminated,
@@ -81,11 +81,11 @@
       Discrimination OtherDiscrimination,
       std::optional<uint16_t> ConstantDiscriminatorOrNone = std::nullopt,
       bool IsIsaPointer = false, bool AuthenticatesNullValues = false)
-      : TheKind(Kind::ARM8_3), IsAddressDiscriminated(IsAddressDiscriminated),
+      : TheKind(llvm::to_underlying(Kind::ARM8_3)), IsAddressDiscriminated(IsAddressDiscriminated),
         IsIsaPointer(IsIsaPointer),
         AuthenticatesNullValues(AuthenticatesNullValues),
-        SelectedAuthenticationMode(AuthenticationMode),
-        DiscriminationKind(OtherDiscrimination), Key(llvm::to_underlying(Key)) {
+        SelectedAuthenticationMode(llvm::to_underlying(AuthenticationMode)),
+        DiscriminationKind(llvm::to_underlying(OtherDiscrimination)), Key(llvm::to_underlying(Key)) {
     assert((getOtherDiscrimination() != Discrimination::Constant ||
             ConstantDiscriminatorOrNone) &&
            "constant discrimination requires a constant!");
@@ -103,7 +103,7 @@
                           OtherDiscrimination, ConstantDiscriminatorOrNone,
                           IsIsaPointer, AuthenticatesNullValues) {}
 
-  Kind getKind() const { return TheKind; }
+  Kind getKind() const { return Kind(TheKind); }
 
   explicit operator bool() const { return isEnabled(); }
 
@@ -130,7 +130,7 @@
 
   Discrimination getOtherDiscrimination() const {
     assert(getKind() != Kind::None);
-    return DiscriminationKind;
+    return Discrimination(DiscriminationKind);
   }
 
   uint16_t getConstantDiscrimination() const {
@@ -149,7 +149,7 @@
   }
 
   PointerAuthenticationMode getAuthenticationMode() const {
-    return SelectedAuthenticationMode;
+    return PointerAuthenticationMode(SelectedAuthenticationMode);
   }
 
   ARM8_3Key getARM8_3Key() const {

```

**Comment by shafik** [comment]
CC @ahmedbougacha @AaronBallman @erichkeane

**Comment by erichkeane** [comment]
Ooof, this is unfortunate.  Typically our support is for the 'latest point release' (so 8.4 and 9.3 are our supported 'versions' of 8 and 9), but I see we also claim support with GCC7.  So does GCC7 have the issue building this?

If so, I think there is value in fixing this, though I'd prefer some sort of use of the PREFERRED_TYPE macro, as well as some way to remind us aggressively to revert this once we stop supporting GCC7.

**Comment by erichkeane** [comment]
Alternatively, we could start the effort to remove GCC7 from our support matrix, its been more than a year (~13 months) since we switched to 7.4: https://github.com/llvm/llvm-project/commit/1e06b82bded69fe627d6cd62ecff236fca15f39b

And the difference between 8.4 and 7.4 is 15 months, so that change might be sensible.

I guess MY PROPOSAL here: 
1- Fix the above in the 19.2 release,a s proposed by the committer
2- Move our support matrix to be GCC 8.4/9.3+ (and perhaps clarify that we mean 'newest point release of each major revision') for Clang 20, so don't change this patch in main.

WDYT @MaskRay @AaronBallman ?

**Comment by jabraham17** [comment]
> So does GCC7 have the issue building this?

Yes, both GCC 7.4 and 7.5 have this issue

**Comment by AaronBallman** [comment]
GCC 7.4 is our minimum required version (https://llvm.org/docs/GettingStarted.html#host-c-toolchain-both-compiler-and-standard-library), so yeah, this should be fixed.

> 1- Fix the above in the 19.2 release,a s proposed by the committer

Agreed.

> 2- Move our support matrix to be GCC 8.4/9.3+ (and perhaps clarify that we mean 'newest point release of each major revision') for Clang 20, so don't change this patch in main.

This requires an RFC and community buy-in. I'm not certain how much pain it would cause to bump to GCC 8. The thread which ended with us [bumping to 7.4](https://discourse.llvm.org/t/require-gcc-7-4-as-gcc-7-3-cannot-build-llvm-16-x-or-main/72310) did not really speak out against 8.x, and from poking around distrowatch, it seems reasonable for LTS releases (Ubuntu 20.04 is on gcc 9.3, Debian 10 (2019) is on gcc 8.3, others within their lifetime seem to be on even new releases).

(Plea: can we start to insist that there is at least one build bot on the minimum version building with -Werror so we can stop regressing our minimum version support? If that means we bump to the oldest version of GCC we have a bot for rather than the oldest version that technically works, that's still better than what we've been doing IMO.)

**Comment by erichkeane** [comment]
> (Plea: can we start to insist that there is at least one build bot on the minimum version building with -Werror so we can stop regressing our minimum version support? If that means we bump to the oldest version of GCC we have a bot for rather than the oldest version that technically works, that's still better than what we've been doing IMO.)

IMO, our policy needs to just be, "We support the ones with working Werror build bots".  Yes, this requires an RFC, but frankly, if the result on that RFC is anything but, "if you care about X old enough compiler, put up a build bot, or shush", i'ma scream.

**Comment by shafik** [comment]
> > (Plea: can we start to insist that there is at least one build bot on the minimum version building with -Werror so we can stop regressing our minimum version support? If that means we bump to the oldest version of GCC we have a bot for rather than the oldest version that technically works, that's still better than what we've been doing IMO.)
> 
> IMO, our policy needs to just be, "We support the ones with working Werror build bots". Yes, this requires an RFC, but frankly, if the result on that RFC is anything but, "if you care about X old enough compiler, put up a build bot, or shush", i'ma scream.

Yeah, if we say we are going to support GCC X then we should have a build bot to catch regressions.

**Comment by joker-eph** [comment]
> Due to a bug in GCC, 

I think this is it to me: I would strongly object making the code **worse** for working around a **warning** in a specific version of GCC. 
Our codebase should be warning-free for recent clang versions, the rest is best effort (patch welcome) and as long as it does not make the code worse (warnings are meant as a tool to make the code better...).

Silencing "buggy" GCC warnings happens, we do it for example: https://github.com/llvm/llvm-project/blob/8db97ae36c586f8e0321b7bc784cda4d8c5bac26/mlir/CMakeLists.txt#L87-L94
Another way to do it can be through pragmas as well: https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html
(I understand it won't help in this case, since somehow gcc didn't expose a flag for this warning...).

Downstream users / plugins can also include LLVM headers as "system include" which will isolate them from warnings in the LLVM codebase, allowing them to build their project with Werror independently.
